### DISCUSSION TITLE
Gradle: Support installation of JDK version requested by Gradle Daemon JVM Discovery

### DISCUSSION THREAD

**Discussion by davidburstrom** [renovatebot/renovate#30164] [Request Help] [answered]
### What would you like help with?

Other

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

The JDK that gets installed in order to run Gradle is resolved by Renovate based on the Gradle version in use [(source)](https://github.com/renovatebot/renovate/blob/198de5843ceae96bb8f3f096bb779a198e3ab985/lib/modules/manager/gradle-wrapper/utils.ts#L51). There is a new Gradle feature, [Daemon JVM discovery](https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:daemon_jvm_criteria), that dictates which (installed) JDK the daemon should be launched with.

By inspecting the `gradle/gradle-daemon-jvm.properties` file (if present), Renovate can resolve which JDK version it should install.

The reason I discovered this was because the job to update the Gradle wrapper (https://github.com/davidburstrom/contester/pull/111) failed to run `./gradlew :wrapper`, and while investigating the job logs, I found that it installed JDK 17 whereas Gradle requested JDK 21.

JDK selection is necessary to support cases where any applied Gradle plugins require a more modern JDK than Gradle itself requires.

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: Using containerbase dynamic installs (branch="renovate/gradle-8.x")
DEBUG: Resolved stable matching version (branch="renovate/gradle-8.x")
{
  "toolName": "java"
  "constraint": "^17.0.0"
  "resolvedVersion": "17.0.11+9"
}

DEBUG: Executing command (branch="renovate/gradle-8.x")
{
  "command": "install-tool java 17.0.11+9"
}

DEBUG: exec completed (branch="renovate/gradle-8.x")
{
  "durationMs": 8002
  "stdout": "[10:50:18.078] INFO (62): Installing tool java@17.0.11+9...\nopenjdk version \"17.0.11\" 2024-04-16\nOpenJDK Runtime Environment Temurin-17.0.11+9 (build 17.0.11+9)\nOpenJDK 64-Bit Server VM Temurin-17.0.11+9 (build 17.0.11+9, mixed mode, sharing)\n[10:50:25.190] INFO (62): Installed tool java in 7.1s.\n"
  "stderr": ""
}

DEBUG: Executing command (branch="renovate/gradle-8.x")
{
  "command": "./gradlew :wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-8.9-bin.zip --gradle-distribution-sha256-sum d725d707bfabd4dfdc958c624003b3c80accc03f7037b5122c4b1d0ef15cecab"
}

DEBUG: rawExec err (branch="renovate/gradle-8.x")
{
  "err": {
    "cmd": "/bin/sh -c ./gradlew :wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-8.9-bin.zip --gradle-distribution-sha256-sum d725d707bfabd4dfdc958c624003b3c80accc03f7037b5122c4b1d0ef15cecab",
    "stderr": "\nFAILURE: Build failed with an exception.\n\n* What went wrong:\nCannot find a Java installation on your machine (Linux 5.15.0-1058-aws amd64) matching the Daemon JVM defined requirements: Compatible with Java 21 (from gradle/gradle-daemon-jvm.properties).\n\n* Try:\n> Run with --stacktrace option to get the stack trace.\n> Run with --info or --debug option to get more log output.\n> Run with --scan to get full insights.\n> Get more help at https://help.gradle.org.\n",```

</details>

**Comment by rarkins** [comment, marked as answer]
Converted to issue: https://github.com/renovatebot/renovate/issues/30165
